### ISSUE TITLE
How to handle window resizing on macOS

### ISSUE THREAD

**Issue by DagAgren** [gfx-rs/wgpu#249]
I am playing around with using wgpu in a macOS project (and eventually iOS, but haven't tried it yet).

Right now I am struggling with how to handle window resizing. I'm currently rendering using CVDisplayLink, and for each frame checking if the size of the window has changed since last frame, and if so, I create a new swap chain.

This is causing double buffer flickering when resizing the window - every other frame is a frozen frame from the double buffer, and ever other is live updating and correct.

If I do not recreate the swap chain at all, it actually looks slightly better: Rendering seems to block completely for short times while resizing, and the contents of the previous frame is just resized instead.

Any clues what is going on? Am I doing something wrong, or is wgpu?

(Also, if I fullscreen, my framerate drops through the floor, and I get occasional squares of corruption on individual frames, but is that the same issue as https://github.com/gfx-rs/wgpu/issues/78 ?)

**Comment by kvark** [comment]
Did you try running https://github.com/gfx-rs/wgpu-rs examples and see how they resize in your environment?

**Comment by DagAgren** [comment]
No, I will try that.

**Comment by DagAgren** [comment]
Sorry, I am not very familiar with rust, so, how do I actually build the examples? "cargo build" does not seem to do it?

**Comment by kvark** [comment]
@DagAgren no worries, we'd be happy to walk you through. Sorry about inconvenience!
Basically, if you install Rust+Cargo (e.g. via https://rustup.rs/), you can do:
```bash
git clone https://github.com/gfx-rs/wgpu-rs
cd wgpu-rs
cargo run --example cube --features metal
```

**Comment by DagAgren** [comment]
Thanks!

Resizing is very broken in a different way in the examples: The animation freezes, and the last frame is stretched to the window size.

**Comment by DagAgren** [comment]
The examples also have the same problem where going to fullscreen will kill the framerate. The animation to fullscreen is of course also frozen the same way.

**Comment by kvark** [comment]
Thank you for verifying! One of the things we are currently doing is stalling the CPU to wait for GPU on resizing: https://github.com/gfx-rs/wgpu/blob/master/wgpu-native/src/device.rs

There is also a number of issues with have with swapchain in gfx-rs Metal backend, including that fullscreen bug. We are looking for ways to resolve them, e.g. https://github.com/gfx-rs/gfx/pull/2882

**Comment by DagAgren** [comment]
All right! I guess in conclusion I should not try too hard to resolve this on my side, but rather wait a while for a few more issues to be addressed and then try again?

**Comment by kvark** [comment]
Yes, that's the way

**Comment by mtak-** [comment]
@DagAgren I would also caution that displaylink runs on a background thread. This could be the cause of the corrupted frames. If you are using objc/swift for CVDisplayLink, I would suggest switching to the [display-link](https://github.com/BrainiumLLC/display-link) crate. It adds a send requirement on the callback which should make any threading issues more obvious.

Also, IIRC winit doesn't support live resize on macos super well until `winit` `0.20` comes out.

**Comment by seivan** [comment]
@mtak- It happens without using `DisplayLink` on macOS as well. Then again pretty use Metal uses one internally.

**Comment by kvark** [comment]
I consider this fixed by #330. Please re-open otherwise!
